### ISSUE TITLE
Light2D is unreliable when reparenting it

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#34673]
**Godot version:**
3.2 beta4

**Issue description:**
Soo I have a rather complicated light setup to workaround #26507 and recently (?) I noticed that it's unreliable. What I do is keep whole game in a ViewportContainer and reparent any Light2D that appears to the parent of this container and then draw the whole screen masked with lights. Yep, that's complicated.

My problem is that lights sometimes don't work. Completely randomly. Like, check this screenshot. The character has light around it.
![image](https://user-images.githubusercontent.com/2223172/71556314-553f4880-2a37-11ea-9da4-72205efacc4c.png)

But when I run the game multiple times, 70% of the time the light does not appear.
![image](https://user-images.githubusercontent.com/2223172/71556353-ec0c0500-2a37-11ea-8557-a4fd747b9756.png)

This happens after reparenting. I tried debugging it to see what's the difference when it works and when doesn't and there doesn't seem to be any. The light has same parent/position in parent in both cases, all properties have the same values in remote inspector. Disabling/enabling light or toggling visibility doesn't fix it. Here's a GIF of the bug in action.
![Ot9elJAW7g](https://user-images.githubusercontent.com/2223172/71556374-34c3be00-2a38-11ea-8a49-be3d0b52f992.gif)

So uh, I have no idea how to reproduce it outside my game, but it's clearly some problem with the engine.

**Minimal reproduction project:**
I haven't managed to create a minimal project :/ Simply reparenting doesn't seem to cause the bug, it's just a trigger.

**Comment by Janders1800** [comment]
Did you set up render priorities on your materials?

**Comment by KoBeWi** [comment]
No, I use default CanvasItemMaterial with render mode set to "Light Only".

**Comment by Beefster09** [comment]
I'm seeing similar issues for lights in instanced scenes inside viewports.

**Comment by Beefster09** [comment]
I'm trying to put together a minimal project without much success.

@KoBeWi does your game use tilemaps at all?

**Comment by KoBeWi** [comment]
It doesn't.

**Comment by Beefster09** [comment]
My game doesn't reparent lights at all, in fact only the scene I'm instancing at runtime is a CanvasLayer that contains a viewport that contains an instanced scene with 2 Light2D nodes, one in add mode and another in mask mode.

It seems the common thread here is viewports and `ViewportContainer`s. There seems to be some weird combination of viewports and lights and potentially another ingredient that makes lights misbehave sometimes.

Is your game's ViewportContainer the descendant of a CanvasLayer?

**Comment by KoBeWi** [comment]
This is my scene tree:
![image](https://user-images.githubusercontent.com/2223172/83133977-f8c67f00-a0e3-11ea-8871-e314c22977e1.png)
My ViewportContainer is only under a Control.

**Comment by Beefster09** [comment]
This is my scene tree:

![unreliable_light_scene_tree](https://user-images.githubusercontent.com/11081313/83158438-02d78600-a0c2-11ea-9e85-cfc3f28c11f2.PNG)

The light is found inside the world_map instanced scene, which looks like this:

![unreliable_light_scene_tree_2](https://user-images.githubusercontent.com/11081313/83158645-43370400-a0c2-11ea-9e62-55c7ff65be81.PNG)

**Comment by xphere** [comment]
Hi there, I have a similar problem and put a minimal project that reproduces it.
In a game I use a light to mask what can be seen, then added a pickup which has its own light. Both scenes worked fine on its own, but when combined, the pickup light dissapears randomly. Even reopening the scene from the editor can trigger this behaviour.

I don't know if it's somehow related to this bug, but I share my minimal project here:
https://github.com/xphere/godot-testing/tree/weird-shadow-behaviour

**Comment by Calinou** [comment]
@KoBeWi @Beefster09 @xphere Can you (or anyone else) still reproduce this bug in [Godot 3.5.1](https://godotengine.org/article/maintenance-release-godot-3-5-1) or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by KoBeWi** [comment]
I no longer use Light2D, so I'm unable to test.

**Comment by xphere** [comment]
Will try to reproduce it later

**Comment by xphere** [comment]
[weird-shadow-behaviour.zip](https://github.com/godotengine/godot/files/9953964/weird-shadow-behaviour.zip)
@Calinou I tried my previous MRP on 3.5.1 and it still behaves erratically.
- 90% of the time the shadow hides the right part of the screen (image A)
- 10% of the time the spotlight is rendered (image B)

Curiously, if I attach a script to reload the scene on keypress, it still behaves weirdly, but the % is more like 50-50. It also happens in GLES3 and GLES2.

Image A:
![image](https://user-images.githubusercontent.com/170968/200373528-f7700510-cf10-40ea-83a4-d4dea19c6257.png)

Image B:
![image](https://user-images.githubusercontent.com/170968/200373726-5566d23e-8af8-4343-815b-38264397d718.png)
